src/olmo_core/nn/hf/checkpoint.py

```diff
@@ -0,0 +1,159 @@
+except ImportError:
+
+    @contextmanager
+    def init_empty_weights(include_buffers: bool = False) -> Generator[None, None, None]:
+        log.warning("accelerate not installed, will initialize weights.")
+        yield None
+
+
+log = logging.getLogger(__name__)
+
+
+@beta_feature
+def load_hf_model(
+    model_name_or_path: PathOrStr,
+    model_state_dict: Dict[str, Any],
+    *,
+    process_group: Optional[dist.ProcessGroup] = None,
+    work_dir: Optional[PathOrStr] = None,
+):
+    """
+    Loads an OLMo Core model state dict using a model in Hugging Face transformers format.
+
+    :param model_name_or_path: The name of a model in HF Hub or the path to a model saved in HF format.
+    :param model_state_dict: The OLMo Core model state dict in which to load HF state.
+    :param process_group: The process group to use for distributed communication.
+    :param work_dir: A local directory that can be used for holding temporary state. Required when
+        downloading a model from a cloud directory.
+    """
+
+    work_dir = f"{work_dir}/hf-tmp" if work_dir is not None else None
+
+    if is_url(model_name_or_path):
+        log.warning(
+            "Model id or path provided is a remote Hugging Face directory. This may not be suitable for unshared file systems."
+        )
+        assert work_dir is not None
+        assert (
+            file_exists(f"{model_name_or_path}/generation_config.json")
+            or file_exists(f"{model_name_or_path}/model.safetensors.index.json")
+            or file_exists(f"{model_name_or_path}/pytorch_model.bin")
+        )
+        model_id = None
+
+        # Download model to local FS
+        if get_fs_local_rank() == 0:
+            copy_dir(model_name_or_path, work_dir)
+        barrier(group=process_group)
+    elif Path(model_name_or_path).is_dir():
+        assert (
+            file_exists(f"{model_name_or_path}/generation_config.json")
+            or file_exists(f"{model_name_or_path}/model.safetensors.index.json")
+            or file_exists(f"{model_name_or_path}/pytorch_model.bin")
+        )
+        model_id = None
+    elif repo_exists(str(model_name_or_path)):
+        log.warning(
+            "Model id or path provided is a Hugging Face model id. This may not be suitable for unshared file systems."
+        )
+        model_id = str(model_name_or_path)
+    else:
+        raise NotImplementedError
+
+    # Warm up the HF local cache by downloading the model on just local rank 0
+    if get_fs_local_rank() == 0:
+        hf_model = AutoModelForCausalLM.from_pretrained(model_name_or_path)
+        del hf_model
+    barrier(group=process_group)
+
+    hf_model = AutoModelForCausalLM.from_pretrained(model_name_or_path)
+    log.info(f"Loaded hf model: {hf_model}")
+
+    converted_state_dict: Dict[str, torch.Tensor] = convert_state_from_hf(
+        hf_model.config, hf_model.state_dict(), model_id=model_id
+    )
+
+    for key, state in converted_state_dict.items():
```

Comment:
Since there's potentially some distributed communication here you'll need to make sure the keys are traversed in the same order on each rank.
```suggestion
    for key in sorted(converted_state_dict.keys()):
        state = converted_state_dict[key]
```